Update references, captions, lit_review and glossary (#4)
larahabashy/capstone-report: capstone-report.md

> By: Ela Bandari, Lara Habashy, Peter Yang and Javairia Raza

Reviewer: Do we want to list alphabetically by last name?

Author: done!

## Executive Summary

> Subclinical lipohypertrophy is traditionally evaluated based on visual inspection or palpation. Recent work has shown that lipohypertrophy may be detected by ultrasound imaging {cite}`kapeluto2018ultrasound`. However, the criteria used to classify lipohypertrophy using ultrasound imaging is only familiar to and implemented by a small group of physicians {cite:ts}`madden_2021`. In an effort to improve the accessibility and efficiency of this method of detection, we have developed a supervised machine learning model to detect lipohypertrophy in ultrasound images. 

Reviewer: ```suggestion
Subclinical lipohypertrophy is traditionally evaluated by visual inspection or palpation. Recent work has shown that lipohypertrophy may be detected by ultrasound imaging {cite}`kapeluto2018ultrasound`. However, the criteria used to classify lipohypertrophy using ultrasound imaging is only familiar to and implemented by a small group of physicians {cite:ts}`madden_2021`. In an effort to improve the accessibility and efficiency of this method of detection, we have developed a supervised machine learning model to detect lipohypertrophy in ultrasound images. 
```

## Data Science Methods 

```{figure} image/ds_workflow.png
---
height: 500px
name: workflow
---
Our general data science workflow for the capstone project. 
```

3.1 Data Preparation

> Given our small dataset, image augmentation techniques are used to expand the size of the training set and ideally make the model more generalizable. We explored the pytorch and the albumentations library for augmentation as they proved to improve accuracy {cite}`al2019deep`. We also attempted to leverage a machine learning tool called autoalbument from the latter library that finds the best transformations for you. However, following several runs, the accuracy did not improve from baseline. We suspect it may be that the transformations were too complex for the model to learn anything significant. Using simpler transformations to augment the data was complemented by results found in our literature review {cite}`esteva2017, loey2020deep`. A variety of classic transformations were tested and the model’s performance on these augmented datasets were documented [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/manual-albumentation.ipynb). The transformations that led to the best performance were adding random vertical and horizontal flipping with a probability of 50% along with random brightness and contrast.

Reviewer: ```suggestion
Given the small size of the dataset, image augmentation techniques were used to expand the size of the training set and improve the model's generalizability. We explored the pytorch and the albumentations libraries as they had proven successful in increasing accuracy in previous work {cite}`al2019deep`. We also attempted to leverage a machine learning tool called autoalbument from the latter library that searches for and selects the best transformations. However, following several runs, the accuracy did not improve from baseline. We suspect it may be that the transformations were too complex for the model to learn anything significant. Using simpler transformations to augment the data was complemented by results found in our literature review {cite}`esteva2017, loey2020deep`. A variety of classic transformations were tested and the model’s performance on these augmented datasets were documented [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/manual-albumentation.ipynb). The transformations that led to the best performance were adding random vertical and horizontal flipping along with random brightness and contrast adjustment, with a probability of 50% .
```

```{figure} image/transformed_image.png
---
height: 200px
name: transformation
---

> Final image transformations included random vertical and horizontal flipping and random brightness and contrast.

Reviewer: ```suggestion
Final image transformations included random vertical and horizontal flipping and random brightness and contrast adjustment.
```

height: 150px
name: acc_chart
---
All four models were tested on a holdout sample to produce these accuracy results. 
```

```{figure} image/model_recall_bar_chart.png
---
height: 150px
name: recall_chart
---
All four models were tested on a holdout sample to produce these recall results. 
```

> The model that utilizes the DenseNet architecture appears to be the best performing model for our dataset. Furthermore, we conducted a [manual analysis](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/model_decision_making.ipynb) of tricky negative and positive images to test model performance using an interactive interface. This application allowed us to compare how confident the models were when misclassifying cases and showcased how well DenseNet was doing. 

Reviewer: ```suggestion
The model that utilizes the DenseNet architecture appears to be the best performing model for our dataset. Furthermore, we conducted a [manual analysis](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/model_decision_making.ipynb) of tricky negative and positive images to evaluate the models' performance using an interactive interface. This application allowed us to compare how confident the models were when misclassifying images and showcased the strengths of the DenseNet architecture . 
```

Reviewer: Do we have a screenshot of DenseNet being the least confident in its wrong prediction that we can add here?

Author: we do! great point :)

> In an effort to reduce the high confidence levels for misclassifications, a problem known as overfitting, we considered adding dropout layers in our CNN structure. In this case, overfitting occurs as a result of having trained our models on such a small training set. Another way to reduce overfitting is to average the predictions from all four models, called an ensemble. However, an ensemble would not be feasible as it requires lots of resources such as enough CPU to train the models. Although the dropout layers did manage to reduce overfitting in the models, the overall reduced accuracy was not significant enough to implement those layers in our optimal model. To see this exploration, click [here](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/densemodels-ax-dropout-layers.ipynb).

Reviewer: Do we have the correct understanding of overfitting here? My understanding is that overfitting is when the model models the training set too closely and not high confidence levels for misclassification. High confidence levels for misclassification may result from overfitting but they may also be due to other errors (in our case wasn't it mostly because we had the wrong test-valid-train set?)

Reviewer: Tagging @larahabashy here who would be the best suited person to address this tomorrow.

Author: @elabandari ! This feedback has been incorporated by @larahabashy and myself!

Furthermore, as our capstone partner was far more interested in reducing false negatives, areas where insulin should not be injected, we investigated various techniques to optimize recall. Positive weight (pos_weight) is one argument of the loss function used in our CNN model which, if greater than 1, prioritizes the positive examples more such that there is a heavier penalization (loss) on labelling the positive lipohypertrophy examples incorrectly. However, this method was not implemented in our final model as it proved to be unstable. After conducting a few experiments, we found high variance in the [results](https://github.com/UBC-MDS/capstone-gdrl-lipo/blob/master/notebooks/pos-weight-exploration.ipynb). 

Finally, the choice of our optimal model as DenseNet was further motivated by its size and compatibility with our data product, further discussed in the data product section.

```{figure} image/size_df.png
---
height: 250px
name: size_df
---
Comparison of the file size of the different models revealed that DenseNet was the smallest model. 
```

> Our next objective was to implement object detection into our pipeline, giving us the ability to identify the exact location of lipohypertrophy on unseen test images. To implement object detection using a popular framework called YOLO, first, using the annotated ultrasound images, the team created bounding boxes around the location of the lipohypertrophy masses in the positive training images. Next, using the YOLOv5 framework, the Yolov5m model was trained for approximately 200 epochs with an image size of 300 and a batch size of 8. The team also experimented with different training parameters to find the optimal ones and found the former specifics optimal.

Reviewer: ```suggestion
Our next objective was to implement object detection into our pipeline, giving us the ability to identify the exact location of lipohypertrophy on unseen test images. To implement object detection using a popular framework called YOLO, the team created bounding boxes around the location of the lipohypertrophy masses in the positive training images using the annotated ultrasound images as a guide. Next, using the YOLOv5 framework, the Yolov5m model was trained for 200 epochs with an image size of 320 and a batch size of 8. The team experimented with different training parameters to find that the aforementioned training parameters produce optimal results.
```

height: 500px
name: obj_det
---

> Our final object detection model results on a test sample reveals promising results. The top row indicates the true location of lipohypertrophy and the bottom row indicates where the model thinks the lipohypertrophy is. The number on the red box indicates the model's confidence. 

Reviewer: ```suggestion
Our final object detection model results on a test sample reveals promising results. The top row indicates the true location of lipohypertrophy and the bottom row indicates the location predicted by the model. The number on the red box indicates the model's confidence. 
```

2. Limited resources - Limited computing power has been a recurrent challenge in this project and has made it difficult to experiment with tuning all the parameters of our model concurrently. 

3. Lack of oversight - We believe that there should be an auditing process involved to ensure that our machine learning model does not propagate any biases that could cause harm to specific patient populations. It would also be useful to have other independent radiologists label the dataset in order to draw comparisons between physicians’ and algorithms’ performance.

To address the limitations outlined above, our team has made the following recommendations: 

> 1.  Addition of more labelled images – In line with the size of the dataset in similar studies {cite}`xiao2018comparison,cheng2017transfer`, we recommend increasing the dataset to at least 1000 images.

Reviewer: ```suggestion
1.  Addition of more labelled images – In line with the size of the dataset in similar studies {cite}`xiao2018comparison,cheng2017transfer`, we recommend increasing the dataset to at least 1000 images and ideally to several thousand images.
```

2.  Increasing computational resources - we recommend obtaining funding for additional computing resources for both the development and deployment of future models.

> 3. Add additional functionality such as native cropping.

Reviewer: ```suggestion
3. Adding additional functionality such as a cropping tool within the web interface.
```
